traffic_ops/testing/api/utils/utils.go

```diff
@@ -73,3 +76,58 @@ func Compare(t *testing.T, expected []string, alertsStrs []string) {
 		}
 	}
 }
+
+type CkReqFunc func(*testing.T, toclientlib.ReqInf, interface{}, error)
+
+func CkRequest(c ...CkReqFunc) []CkReqFunc {
+	return c
+}
+
+func NoError() CkReqFunc {
+	return func(t *testing.T, _ toclientlib.ReqInf, _ interface{}, err error) {
+		assert.NoError(t, err, "Expected no error, but got %v", err)
+	}
+}
+
+func HasError() CkReqFunc {
+	return func(t *testing.T, _ toclientlib.ReqInf, resp interface{}, err error) {
+		switch resp := resp.(type) {
+		case tc.CacheGroupDetailResponse:
+			assert.Error(t, err, "Expected error - but got: %+v %+v", resp.Response, resp.Alerts)
+		case tc.CacheGroupsNullableResponse:
+			assert.Error(t, err, "Expected error - but got: %+v %+v", resp.Response, resp.Alerts)
+		case tc.Alerts:
+			assert.Error(t, err, "Expected error - but got: %+v", resp.Alerts)
+		default:
+			assert.Fail(t, "Failed to find response type")
+		}
+	}
+}
+
+func HasStatus(expectedStatus int) CkReqFunc {
+	return func(t *testing.T, reqInf toclientlib.ReqInf, _ interface{}, _ error) {
+		assert.Equal(t, expectedStatus, reqInf.StatusCode, "Expected Status Code %d, got %d", expectedStatus, reqInf.StatusCode)
+	}
+}
+
+func ResponseHasLength(expected int) CkReqFunc {
+	return func(t *testing.T, _ toclientlib.ReqInf, resp interface{}, _ error) {
+		switch resp := resp.(type) {
+		case tc.CacheGroupsNullableResponse:
+			assert.Equal(t, expected, len(resp.Response), "Expected response object length %d, but got %d", expected, len(resp.Response))
+		default:
+			assert.Fail(t, "Failed to find response type")
+		}
+	}
+}
```

Comment:
Same as above RE: how massive this will get if it has to switch through everything.

If this were to take in a `[]interface{}` that would make it much simpler, but it would cost the ability to to use it generically in structures and call signatures that call for a `CkReqFunc`. Which wouldn't be a terrible thing, in my opinion, but it would be a lot of work to change and it'd make the test cases a bit harder to write, not as nice as you have below that's for sure.

So instead if the Response and alerts are passed in separately in `CkReqFunc`s as I suggested above, you could narrow the type of the response to a slice by just doing
```go
func ResponseHasLength(expected int) CkReqFunc {
	return func(t *testing.T, _ toclientlib.ReqInf, resp interface{}, _ tc.Alerts, _ error) {
		if slice, ok := resp.([]interface{}); ok {
			assert.Equal(t, expected, len(slice), "Expected response object length %d, but got %d", expected, len(slice))
		} else {
			t.Failf("Expected response to be an array of length %d, but it wasn't an array at all: %+v", expected, resp)
		}
	}
}
```

Reply:
Also same as above RE: error vs failure

Reply:
Type cast didn't work but reflection did

Reply:
that's strange, type casting worked for me in a playground